sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala

```diff
@@ -202,8 +202,11 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
     // We can only parse the value as TimestampNTZType if it does not have zone-offset or
     // time-zone component and can be parsed with the timestamp formatter.
     // Otherwise, it is likely to be a timestamp with timezone.
-    if (timestampNTZFormatter.parseWithoutTimeZoneOptional(field, false).isDefined) {
-      SQLConf.get.timestampType
+    val timestampType = SQLConf.get.timestampType
+    if ((SQLConf.get.legacyTimeParserPolicy == LegacyBehaviorPolicy.LEGACY ||
+        timestampType == TimestampNTZType) &&
+        timestampNTZFormatter.parseWithoutTimeZoneOptional(field, false).isDefined) {
```

Comment:
Could you clarify this:
(legacyTimeParserPolicy = LEGACY || timestampType == TimestampLTZType)
we are trying to parse it as NTZ, and it is parsable we return `TimestampLTZType`?

This confuses me, return `TIMESTAMP LTZ` when the input was parsed by a NTZ function.

Reply:
> Because the LEGACY behavior used timestampNTZFormatter to parse timestamp.

I see. It is ok if the such legacy behaviour is covered by a test.

Reply:
> Because the LEGACY behavior used timestampNTZFormatter to parse timestamp. 

I can't find the related code, @Hisoka-X can you point to it?

Reply:
I did some test use old version which before this PR merged. When `SQLConf.get.timestampType == TIMESTAMP_LTZ`, it used `timestampNTZFormatter` to parsed timestamp and passed. You can use `CSVLegacyTimeParserSuite.SPARK-37326: Timestamp type inference for a column with TIMESTAMP_NTZ values` to reproduce this behavior.
![image](https://github.com/apache/spark/assets/32387433/619c7ba2-3cf2-4938-b9a4-4f5079be9421)

Reply:
so the condition here should just be `timestampType == TimestampNTZType`?

Reply:
The purpose of adding `legacyTimeParserPolicy == LEGACY` is to keep the legacy behavior unchanged. But legacy’s behavior is indeed not correct.

Reply:
My point is we should fully understand the code before making changes. Why do we ever use ntz formatter to infer ltz type?

Reply:
Because the string format are same of two type.

Reply:
Then we hit the else branch and `tryParseTimestamp` can infer the type properly?

Reply:
According test case, I think yes. Is any case not right now?

Reply:
The code looks wrong, we may infer ltz using the nzt formatter. This can be a potential bug and bite us in the future.

Reply:
Yep, but it only happened when use legacy mode. Feel free to change it if you think the legacy behavior not right.